rsocket-core/src/test/java/io/rsocket/SetupRejectionTest.java

```diff
@@ -0,0 +1,148 @@
+import reactor.test.StepVerifier;
+
+public class SetupRejectionTest {
+
+  @Test
+  void responderRejectSetup() {
+    SingleConnectionTransport transport = new SingleConnectionTransport();
+
+    String errorMsg = "error";
+    RejectingAcceptor acceptor = new RejectingAcceptor(errorMsg);
+    RSocketFactory.receive().acceptor(acceptor).transport(transport).start().block();
+
+    transport.connect();
+
+    Frame sentFrame = transport.awaitSent();
+    assertThat(sentFrame.getType()).isEqualTo(FrameType.ERROR);
+    RuntimeException error = Exceptions.from(sentFrame);
+    assertThat(errorMsg).isEqualTo(error.getMessage());
+    assertThat(error).isInstanceOf(RejectedSetupException.class);
+    RSocket acceptorSender = acceptor.senderRSocket().block();
+    assertThat(acceptorSender.isDisposed()).isTrue();
+  }
+
+  @Test
+  void requesterStreamsTerminatedOnZeroErrorFrame() {
+    TestDuplexConnection conn = new TestDuplexConnection();
+    List<Throwable> errors = new ArrayList<>();
+    RSocketClient rSocket =
+        new RSocketClient(
+            conn, DefaultPayload::create, errors::add, StreamIdSupplier.clientSupplier());
+
+    String errorMsg = "error";
+
+    Mono.delay(Duration.ofMillis(100))
+        .doOnTerminate(
+            () ->
+                conn.addToReceivedBuffer(Frame.Error.from(0, new RejectedSetupException(errorMsg))))
+        .subscribe();
+
+    StepVerifier.create(rSocket.requestResponse(DefaultPayload.create("test")))
+        .expectErrorMatches(
+            err -> err instanceof RejectedSetupException && errorMsg.equals(err.getMessage()))
+        .verify(Duration.ofSeconds(5));
+
+    assertThat(errors).hasSize(1);
+    assertThat(rSocket.isDisposed()).isTrue();
+  }
+
+  @Test
+  void requesterNewStreamsTerminatedAfterZeroErrorFrame() {
+    TestDuplexConnection conn = new TestDuplexConnection();
+    RSocketClient rSocket =
+        new RSocketClient(
+            conn, DefaultPayload::create, err -> {}, StreamIdSupplier.clientSupplier());
+
+    conn.addToReceivedBuffer(Frame.Error.from(0, new RejectedSetupException("error")));
+
+    StepVerifier.create(
+            rSocket
+                .requestResponse(DefaultPayload.create("test"))
+                .delaySubscription(Duration.ofMillis(100)))
+        .expectErrorMatches(
+            err -> err instanceof RejectedSetupException && "error".equals(err.getMessage()))
+        .verify(Duration.ofSeconds(5));
+  }
+
+  private static class RejectingAcceptor implements SocketAcceptor {
+    private final String errorMessage;
+
+    public RejectingAcceptor(String errorMessage) {
+      this.errorMessage = errorMessage;
+    }
+
+    private final UnicastProcessor<RSocket> senderRSockets = UnicastProcessor.create();
+
+    @Override
+    public Mono<RSocket> accept(ConnectionSetupPayload setup, RSocket sendingSocket) {
+      senderRSockets.onNext(sendingSocket);
+      return Mono.error(new RuntimeException(errorMessage));
```

Comment:
Why not use RejectedSetupException?

Reply:
I thought It is not feasible to force (and expect) people use correct exception, so this test verifies any exception is translated  to RejectedSetupException